sql/dd/impl/upgrade/server.cc

```diff
@@ -788,6 +788,23 @@ bool do_server_upgrade_checks(THD *thd) {
     }
   }
 
+  if (!error_count.has_too_many_errors()) {
+    /*
+      Get hold of the InnoDB handlerton. The check for partitioned tables
+      using shared tablespaces is only relevant for InnoDB.
+    */
+    plugin_ref pr =
+        ha_resolve_by_name_raw(thd, LEX_CSTRING{STRING_WITH_LEN("InnoDB")});
+    handlerton *hton =
+        (pr != nullptr ? plugin_data<handlerton *>(pr) : nullptr);
+    DBUG_ASSERT(hton != nullptr && hton->is_there_mk_to_keyring_rotation);
+
+    if (hton->is_there_mk_to_keyring_rotation()) {
```

Comment:
Why do we need this new handlerton API? Because when tablespace are loaded, InnoDB initialization would fail?
(for 5.7->8.0 and 8.0.X->8.0.Y upgrades)

The handlerton API that uses for tablespace upgrade should be reused for this purpose.

Reply:
As we have discussed on slack. When server is initialized it validates page0 of each tablespace. This is were we can check if tablespace is in rotation from MK encryption to keyring encryption for free. This handleron API indicates whether such table was discovered during the startup. It is checked in do_server_upgrade_check - which is natural place where to check for situations that would cause the upgrade to fail.